### PR TITLE
Retry op when getting 401 status from Sam

### PR DESCRIPTION
**Pull request by aweng98**
\<your comments for this PR go here\>

Have you read [CONTRIBUTING.md](../CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've documented my API changes in Swagger

In all cases:

- [ ] Get two thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; you can delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/sam/SamAuthProvider.scala [hunk 1]
```diff
@@ -127,11 +129,15 @@ class SamAuthProvider(val config: Config, serviceAccountProvider: ServiceAccount
...
+      owningProjects <- retryUntilSuccessOrTimeout(shouldInvalidateSamCacheAndRetry, s"SamAuthProvider.owningProjects.filterUserVisibleClusters call failed")(samRetryInterval, samRetryTimeout) { () =>
+        Future(blocking(samClient.listOwningProjects(userInfo)).toSet)
+      }
+      createdClusters <- retryUntilSuccessOrTimeout(shouldInvalidateSamCacheAndRetry, s"SamAuthProvider.owningProjects.filterUserVisibleClusters call failed")(samRetryInterval, samRetryTimeout) { () =>
+        Future(blocking(samClient.listCreatedClusters(userInfo)).toSet)
+      }
     } yield {
       clusters.filter { case (project, name) =>
-        owningProjects.contains(project) || createdClusters.contains((project, name))
+        owningProjects.contains(project) || createdClusters.right.get._2.contains((project, name))
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/sam/SamAuthProvider.scala]: I don't think `.right.get._2` is correct in this case. `retryUntilSuccessOrTimeout` returns a `RetryableFuture`. What you _want_ is this implicit conversion to kick in: https://github.com/broadinstitute/workbench-libs/blob/develop/util/src/main/scala/org/broadinstitute/dsde/workbench/util/Retry.scala#L98

However I don't think it is in this case. You may need to call it explicitly. This code might work:
```
  override def filterUserVisibleClusters(userInfo: UserInfo, clusters: List[(GoogleProject, ClusterName)])(implicit executionContext: ExecutionContext): Future[List[(GoogleProject, ClusterName)]] = {
    for {
      owningProjects <- retryableFutureToFuture(retryUntilSuccessOrTimeout(shouldInvalidateSamCacheAndRetry, s"SamAuthProvider.owningProjects.filterUserVisibleClusters call failed")(samRetryInterval, samRetryTimeout) { () =>
        Future(blocking(samClient.listOwningProjects(userInfo)).toSet)
      })
      createdClusters <- retryableFutureToFuture(retryUntilSuccessOrTimeout(shouldInvalidateSamCacheAndRetry, s"SamAuthProvider.owningProjects.filterUserVisibleClusters call failed")(samRetryInterval, samRetryTimeout) { () =>
        Future(blocking(samClient.listCreatedClusters(userInfo)).toSet)
      })
    } yield {
      clusters.filter { case (project, name) =>
        owningProjects.contains(project) || createdClusters.contains((project, name))
      }
    }
  }
```

**Comment by rtitle** [comment]
One comment above -- the unit test failure is probably due to the issue in `filterUserVisibleClusters`.

**Comment by aweng98** [comment]
jenkins retest

**Comment by aweng98** [comment]
docker images pull failed.

jenkins retest

**Comment by rtitle** [comment]
I think you missed one: `SamAuthProvider.hasProjectPermission()`. Otherwise looks good.

**Comment by aweng98** [comment]
eagle-eyed! added missing retry call in that method.

**Comment by aweng98** [comment]
one related question. retry timeout seems to be really small at 200 millsec. `samRetryTimeout = 200 milliseconds`. Do you agree with increase it a little?

**Comment by rtitle** [comment]
Hm yeah it is pretty short. I guess currently it tries up to 3 times, sleeping 100ms between each try. So:

call <fail>
sleep(100)
call <fail>
sleep(100)
call <fail>
exception is thrown

I think we didn't want it to retry so long that the actual _Leo_ API call would time out -- then again 100ms is pretty short. I bet retrying 3 times sleeping 500ms or even 1 second would be fine.

**Pull request merged**